src/Microsoft.AspNet.TestHost/TestServer.cs

```diff
@@ -14,18 +14,20 @@
 
 namespace Microsoft.AspNet.TestHost
 {
-    public class TestServer : IServerFactory, IDisposable
+    public class TestServer : IDisposable
     {
         private const string DefaultEnvironmentName = "Development";
         private const string ServerName = nameof(TestServer);
-        private static readonly IFeatureCollection ServerInfo = new FeatureCollection();
-        private Func<IFeatureCollection, Task> _appDelegate;
+        private RequestDelegate _appDelegate;
         private IDisposable _appInstance;
         private bool _disposed = false;
+        private IHttpContextAccessor _httpContextAccessor;
 
         public TestServer(WebHostBuilder builder)
         {
-            _appInstance = builder.UseServer(this).Build().Start();
+            var hostingEngine = builder.UseServer(new ServerFactory(requestDelegate => { _appDelegate = requestDelegate; })).Build();
```

Comment:
I can. What's the order of precedence here? Look for the server, if not found try to create one with the resolved serverfactory?